File: src/game/cosmetics/wardrobe.ts
import type { GameDatabase } from '../data/types'
import type { PlayerAppearance, PlayerSave } from '../save/types'
import {
  APPEARANCE_CATEGORIES,
  appearanceCategoryLabel,
  appearanceOptions,
  type AppearanceCategory,
} from './appearance'
import {
  cosmeticById,
  cosmeticSlotById,
  cosmeticSlots,
  cosmeticsForSlot,
  equippedCosmeticId,
  isCosmeticUnlocked,
} from './cosmetics'

function itemName(db: GameDatabase, itemId: string): string | null {
  return db.Items.find((row) => row['Item ID'] === itemId)?.['Display Name'] ?? null
}

/** One tab of the wardrobe, in table order. */
export interface WardrobeSlotTab {
  slotId: string
  label: string
}

export function wardrobeSlotTabs(db: GameDatabase): WardrobeSlotTab[] {
  return cosmeticSlots(db).map((slot) => ({
    slotId: slot['Cosmetic Slot ID'],
    label: slot['Display Name'],
  }))
}

/** One owned cosmetic, as its tile reads. */
export interface WardrobeTile {
  cosmeticId: string
  itemId: string
  name: string
  equipped: boolean
}

/** The wardrobe's lower half for one slot: what is owned and what is worn. */
export interface WardrobeSlotView {
  slotId: string
  label: string
  equippedCosmeticId: string | null
  /** Only what the player owns — a wardrobe is not a catalogue. */
  tiles: WardrobeTile[]
  /** What to say instead of tiles when the slot has nothing in it yet. */
  emptyNote: string
}

/**
 * The wardrobe's view of one slot.
 *
 * Falls back to the first slot when [slotId] names none, so a client can open
 * on a remembered tab without checking whether it still exists.
 */
export function wardrobeSlotView(
  db: GameDatabase,
  save: PlayerSave,
  slotId: string,
): WardrobeSlotView | null {
  const slot = cosmeticSlotById(db, slotId) ?? cosmeticSlots(db)[0]
  if (!slot) return null
  const resolvedId = slot['Cosmetic Slot ID']
  const equipped = equippedCosmeticId(save, resolvedId)

  return {
    slotId: resolvedId,
    label: slot['Display Name'],
    equippedCosmeticId: equipped,
    tiles: cosmeticsForSlot(db, resolvedId)
      .filter((cosmetic) => isCosmeticUnlocked(save, cosmetic['Cosmetic ID']))
      .map((cosmetic) => {
        const cosmeticId = cosmetic['Cosmetic ID']
        const itemId = cosmetic['Item ID']
        return {
          cosmeticId,
          itemId,
          name: itemName(db, itemId) ?? cosmeticId,
          equipped: cosmeticId === equipped,
        }
      }),
    emptyNote: `No ${slot['Display Name']} unlocked yet.`,
  }
}

/** One appearance row: a label and the stops its slider can land on. */
export interface AppearanceSlider {
  category: AppearanceCategory
  label: string
  optionIds: string[]
  /** Index into [optionIds]; 0 when the save holds an option the table dropped. */
  selectedIndex: number
}

/**
 * The appearance rows, in category order, skipping any the content has no
 * options for.
 *
 * Takes the appearance rather than the save because character creation picks a
 * look before there is a save to put it in.
 */
export function appearanceSliders(
  db: GameDatabase,
  appearance: PlayerAppearance,
): AppearanceSlider[] {
  const sliders: AppearanceSlider[] = []
  for (const category of APPEARANCE_CATEGORIES) {
    const optionIds = appearanceOptions(db, category).map((row) => row['Appearance Option ID'])
    if (optionIds.length === 0) continue
    sliders.push({
      category,
      label: appearanceCategoryLabel(category),
      optionIds,
      selectedIndex: Math.max(0, optionIds.indexOf(appearance[category])),
    })
  }
  return sliders
}

/** What the popup says when a cosmetic is unlocked. */
export interface CosmeticUnlockNotice {
  cosmeticId: string
  /** Null when the cosmetic has no item behind it, which only bad data causes. */
  itemId: string | null
  name: string
  /** The one-time "tap your portrait" line, for the player's first cosmetic. */
  hint: string | null
}

const WARDROBE_HINT =
  'Tap your portrait in the top-left corner anytime to open the Wardrobe and equip it.'

export function cosmeticUnlockNotice(
  db: GameDatabase,
  cosmeticId: string,
  isFirstEver: boolean,
): CosmeticUnlockNotice | null {
  const cosmetic = cosmeticById(db, cosmeticId)
  if (!cosmetic) return null
  const itemId = cosmetic['Item ID']
  return {
    cosmeticId,
    itemId: itemId ?? null,
    name: (itemId ? itemName(db, itemId) : null) ?? cosmeticId,
    hint: isFirstEver ? WARDROBE_HINT : null,
  }
}

File: packages/ik_rules/lib/src/cosmetics/wardrobe.dart
import 'package:collection/collection.dart';
import 'package:ik_content/ik_content.dart';

import '../save/generated/save_models.dart';
import 'appearance.dart';
import 'cosmetics.dart';

String? _itemName(GameDatabase db, String itemId) {
  final displayName = db.items
      .firstWhereOrNull((row) => row.raw['Item ID'] == itemId)
      ?.raw['Display Name'];
  return displayName is String ? displayName : null;
}

String _slotLabel(CosmeticSlotRow slot) {
  final label = slot.raw['Display Name'];
  return label is String ? label : slot.raw['Cosmetic Slot ID'] as String;
}

/// One tab of the wardrobe, in table order.
class WardrobeSlotTab {
  const WardrobeSlotTab({required this.slotId, required this.label});

  final String slotId;
  final String label;

  Map<String, Object?> toJson() => <String, Object?>{'slotId': slotId, 'label': label};
}

List<WardrobeSlotTab> wardrobeSlotTabs(GameDatabase db) {
  return cosmeticSlots(db)
      .map(
        (slot) => WardrobeSlotTab(
          slotId: slot.raw['Cosmetic Slot ID'] as String,
          label: _slotLabel(slot),
        ),
      )
      .toList();
}

/// One owned cosmetic, as its tile reads.
class WardrobeTile {
  const WardrobeTile({
    required this.cosmeticId,
    required this.itemId,
    required this.name,
    required this.equipped,
  });

  final String cosmeticId;
  final String itemId;
  final String name;
  final bool equipped;

  Map<String, Object?> toJson() => <String, Object?>{
    'cosmeticId': cosmeticId,
    'itemId': itemId,
    'name': name,
    'equipped': equipped,
  };
}

/// The wardrobe's lower half for one slot: what is owned and what is worn.
class WardrobeSlotView {
  const WardrobeSlotView({
    required this.slotId,
    required this.label,
    required this.equippedCosmeticId,
    required this.tiles,
    required this.emptyNote,
  });

  final String slotId;
  final String label;
  final String? equippedCosmeticId;

  /// Only what the player owns — a wardrobe is not a catalogue.
  final List<WardrobeTile> tiles;

  /// What to say instead of tiles when the slot has nothing in it yet.
  final String emptyNote;

  Map<String, Object?> toJson() => <String, Object?>{
    'slotId': slotId,
    'label': label,
    'equippedCosmeticId': equippedCosmeticId,
    'tiles': tiles.map((tile) => tile.toJson()).toList(),
    'emptyNote': emptyNote,
  };
}

/// The wardrobe's view of one slot.
///
/// Falls back to the first slot when [slotId] names none, so a client can open
/// on a remembered tab without checking whether it still exists.
WardrobeSlotView? wardrobeSlotView(GameDatabase db, PlayerSave save, String slotId) {
  final slot = cosmeticSlotById(db, slotId) ?? cosmeticSlots(db).firstOrNull;
  if (slot == null) return null;
  final resolvedId = slot.raw['Cosmetic Slot ID'] as String;
  final equipped = equippedCosmeticId(save, resolvedId);

  return WardrobeSlotView(
    slotId: resolvedId,
    label: _slotLabel(slot),
    equippedCosmeticId: equipped,
    tiles: cosmeticsForSlot(db, resolvedId)
        .where((cosmetic) => isCosmeticUnlocked(save, cosmetic.raw['Cosmetic ID'] as String))
        .map((cosmetic) {
          final cosmeticId = cosmetic.raw['Cosmetic ID'] as String;
          final itemId = cosmetic.raw['Item ID'] as String;
          return WardrobeTile(
            cosmeticId: cosmeticId,
            itemId: itemId,
            name: _itemName(db, itemId) ?? cosmeticId,
            equipped: cosmeticId == equipped,
          );
        })
        .toList(),
    emptyNote: 'No ${_slotLabel(slot)} unlocked yet.',
  );
}

/// One appearance row: a label and the stops its slider can land on.
class AppearanceSlider {
  const AppearanceSlider({
    required this.category,
    required this.label,
    required this.optionIds,
    required this.selectedIndex,
  });

  final AppearanceCategory category;
  final String label;
  final List<String> optionIds;

  /// Index into [optionIds]; 0 when the save holds an option the table dropped.
  final int selectedIndex;

  Map<String, Object?> toJson() => <String, Object?>{
    'category': category.key,
    'label': label,
    'optionIds': optionIds,
    'selectedIndex': selectedIndex,
  };
}

/// The appearance rows, in category order, skipping any the content has no
/// options for.
///
/// Takes the appearance rather than the save because character creation picks a
/// look before there is a save to put it in.
List<AppearanceSlider> appearanceSliders(GameDatabase db, PlayerAppearance appearance) {
  final sliders = <AppearanceSlider>[];
  for (final category in AppearanceCategory.values) {
    final optionIds = appearanceOptions(
      db,
      category,
    ).map((row) => row.raw['Appearance Option ID'] as String).toList();
    if (optionIds.isEmpty) continue;
    final index = optionIds.indexOf(appearanceSelection(appearance, category));
    sliders.add(
      AppearanceSlider(
        category: category,
        label: appearanceCategoryLabel(category),
        optionIds: optionIds,
        selectedIndex: index < 0 ? 0 : index,
      ),
    );
  }
  return sliders;
}

/// What the popup says when a cosmetic is unlocked.
class CosmeticUnlockNotice {
  const CosmeticUnlockNotice({
    required this.cosmeticId,
    required this.itemId,
    required this.name,
    required this.hint,
  });

  final String cosmeticId;

  /// Null when the cosmetic has no item behind it, which only bad data causes.
  final String? itemId;
  final String name;

  /// The one-time "tap your portrait" line, for the player's first cosmetic.
  final String? hint;

  Map<String, Object?> toJson() => <String, Object?>{
    'cosmeticId': cosmeticId,
    'itemId': itemId,
    'name': name,
    'hint': hint,
  };
}

const String _wardrobeHint =
    'Tap your portrait in the top-left corner anytime to open the Wardrobe and equip it.';

CosmeticUnlockNotice? cosmeticUnlockNotice(GameDatabase db, String cosmeticId, bool isFirstEver) {
  final cosmetic = cosmeticById(db, cosmeticId);
  if (cosmetic == null) return null;
  final itemId = cosmetic.raw['Item ID'];
  return CosmeticUnlockNotice(
    cosmeticId: cosmeticId,
    itemId: itemId is String ? itemId : null,
    name: (itemId is String ? _itemName(db, itemId) : null) ?? cosmeticId,
    hint: isFirstEver ? _wardrobeHint : null,
  );
}
